wrong result of integer division in presence of assert within conditional operator expression

## kdevel reported this on 2022-05-19T14:54:18Z

### Transferred from https://issues.dlang.org/show_bug.cgi?id=23126

### CC List

* basile-z
* Siarhei Siamashka
* Paul Backus

### Description

````markdown
Found the function divide in a forum post [1]

```div.d
import std.stdio;
import std.conv;

int divide(int dividend, int divisor)
{
    return divisor == 0
        ? assert(false, "You shall not divide by zero")
        : dividend / divisor;
}

int main (string [] args)
{
   int a = args[1].to!int;
   int b = args[2].to!int;
   int c = divide (a, b);
   writefln!"%d / %d = %d" (a, b, c);
   return 0;
}
```

$ dmd div.d
$ ./div 1 2
1 / 2 = 2

Wrong result since dmd 2.097.2, dmd 2.096.1 complains

   div.d(6): Error: `return` expression expected

dmd -O produces executable returning correct results.

expected:

$ gdc div.d -o div
$ ./div 1 2
1 / 2 = 0

[1] <http://forum.dlang.org/thread/vfkwhtigkvkmhtzalwaf@forum.dlang.org?page=3#post-ywralegnwtfiztrkzoak:40forum.dlang.org>
````

#### kdevel commented on 2023-12-30T11:46:29Z



````markdown
(In reply to Siarhei Siamashka from comment #3)
> (In reply to kdevel from comment #2)
> > Works since dmd 2.101.1.
> 
> Does the DMD test suite have a test for it? To ensure that this problem does
> not re-appear again in the future.
Simple question, not easy to answer. For now more than two hours I am trying to bisect dmd between v2.100.2 and v2.101.1 to no avail. I have stopped when git asked me to compile a commit of 2016

[e8ab5ca21f3010bb40509ca7aaf575440886795b] Merge pull request #1639 from WalterBright/rm-PostBlitType

$ make -f posix.mak 
../dmd/src/dmd -conf= -c -o- -Isrc -Iimport -Hfimport/core/sync/barrier.di src/core/sync/barrier.d
make: ../dmd/src/dmd: Command not found
make: *** [import/core/sync/barrier.di] Error 127

which fails. I could not git bisect skip this often enough in a reasonable amount of time.

dmd 2.105.3 could not compile the old compiler:

[...]
(DC) COMMON

The following operation failed:
Name: common
Sources: 
 -> /tmp/k/dmd/src/dmd/common/bitfields.d
 -> /tmp/k/dmd/src/dmd/common/file.d
 -> /tmp/k/dmd/src/dmd/common/int128.d
 -> /tmp/k/dmd/src/dmd/common/outbuffer.d
 -> /tmp/k/dmd/src/dmd/common/string.d

Targets: 
 -> /tmp/k/dmd/generated/linux/release/64/common.o

Command: [...]dmd2/linux/bin64/dmd -c -of/tmp/k/dmd/generated/linux/release/64/common.o -version=MARS -w -de -fPIC -m64 -J/tmp/k/dmd/generated/linux/release/64 -I/tmp/k/dmd/src -dip25 -g -color=on src/dmd/common/bitfields.d src/dmd/common/file.d src/dmd/common/int128.d src/dmd/common/outbuffer.d src/dmd/common/string.d

-----------------------------------------------------------
Deprecation: `-dip25` no longer has any effect

make[1]: *** [dmd] Error 1

Therefor I used dmd 2.098.1 as HOST_DMD
````

--- a/compiler/test/runnable/noreturn2.d
+++ b/compiler/test/runnable/noreturn2.d
@@ -273,11 +273,26 @@ void testCast()
     assert(0);
 }
 
+// https://github.com/dlang/dmd/issues/20108
+int divide(int dividend, int divisor)
+{
+    return divisor == 0
+        ? assert(0, "You shall not divide by zero")
+        : dividend / divisor;
+}
+
+void testConditional()
+{
+    assert(divide(1, 2) == 0);
+    assert(divide(6, 2) == 3);
+}
+
 int main()
 {
     testDtors();
     testAccess();
     testFuncCall();
     testCast();
+    testConditional();
     return 0;
 }

